homeassistant/components/s3/backup.py

```diff
@@ -0,0 +1,218 @@
+from . import S3ConfigEntry
+from .const import CONF_BUCKET, DATA_BACKUP_AGENT_LISTENERS, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+METADATA_VERSION = "1"
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    entries: list[S3ConfigEntry] = hass.config_entries.async_loaded_entries(DOMAIN)
+    return [S3BackupAgent(hass, entry) for entry in entries]
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+
+    return remove_listener
+
+
+def _get_key(backup: AgentBackup) -> str:
+    """Return the key for a backup."""
+    return f"{backup.backup_id}_{suggested_filename(backup)}"
+
+
+def _serialize(metadata: dict[str, Any]) -> dict[str, str]:
+    """Serialize metadata."""
+    return {k: json.dumps(v) for k, v in metadata.items()}
+
+
+def _deserialize(metadata: dict[str, str]) -> dict[str, Any]:
+    """Deserialize metadata."""
+    return {k: json.loads(v) for k, v in metadata.items()}
+
+
+class S3BackupAgent(BackupAgent):
+    """Backup agent for the S3 integration."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: S3ConfigEntry) -> None:
+        """Initialize the S3 agent."""
+        super().__init__()
+        self._client = entry.runtime_data
+        self._bucket: str = entry.data[CONF_BUCKET]
+        self.name = entry.title
+        self.unique_id = entry.entry_id
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :return: An async iterator that yields bytes.
+        """
+        _, obj = await self._get_backup(backup_id)
+        response = await self._client.get_object(Bucket=self._bucket, Key=obj["Key"])
+        return response["Body"].iter_chunks()
+
+    async def async_upload_backup(
+        self,
+        *,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        backup: AgentBackup,
+        **kwargs: Any,
+    ) -> None:
+        """Upload a backup.
+
+        :param open_stream: A function returning an async iterator that yields bytes.
+        :param backup: Metadata about the backup that should be uploaded.
+        """
+        key = _get_key(backup)
+        metadata = {
+            "metadata_version": METADATA_VERSION,
+            "backup_metadata": backup.as_dict(),
```

Comment:
```suggestion
            "backup_metadata": json.dumps(backup.as_dict()),
```
why not just that instead of the serialize method? would be more readable